mir: separate NaN checks from FP arithmetic

zerbina opened the pull request:
## Summary

Decouple NaN checks from floating point arithmetic by emitting a
dedicated magic call during AST -> MIR translation.

This makes the exceptional control-flow arising from NaN checks visible
at the MIR level, which fixes locals not being destroyed if their scope
is left via an exception raised by a NaN check and there's no other
unstructured control-flow leaving the scope.

## Details

* if NaN checks are enabled, `mirgen` emits a call to the `mChckNaN`
  magic after each FP arithmetic operation (add, sub, mul, div)

* `mChckNaN` is a new internal-only unary magic

* `cgen` implements the magic in the same way it previously implemented
  NaN checks; NaN checks are no longer emitted manually for FP
  arithmetic

* `jsgen` and `vmgen` treat the `mChckNaN` as a no-op for now (both
  were already missing NaN check support previously)

zerbina committed:
mir: split NaN checks from FP arithmetic

`mirgen` now emits an explicit NaN check for floating-point arithmetic,
making the exceptional control-flow explicit and visible to
data-/control-flow analysis.

The NaN check is represented by the new `mChckNaN` internal-only magic.

zerbina committed:
cgen: implement the `mChckNaN` magic

The magic is implemented in the same way that nan-checks were so far
implemented by `cgen`, namely, by a nan-check + `raiseFloatInvalidOp`
call.

A comment about potentially undesired behaviour with this
implementation regarding floating point exceptions is also added.

zerbina committed:
jsgen, vmgen: treat the `mChckNaN` as a no-op

This is in line with how nan-checks previously worked with both
backends (read: not at all). They need to be implemented eventually,
but this is left to future work.

zerbina committed:
tests: ensure proper destruction when a NaN check fails

Since the exceptional control-flow of NaN checks wasn't visible,
destruction of locals didn't work in all cases. Now it does.

saem commented on compiler/backend/ccgexprs.nim:
```diff
@@ -1763,6 +1760,15 @@ proc genMagicExpr(p: BProc, e: CgNode, d: var TLoc, op: TMagic) =
+    initLocExpr(p, e[1], a)
+    # XXX: using a comparison could result in a floating point exception to
+    #      be raised (i.e., the flag to be set in the fenv).
+    #      ``__builtin_isnan``, which doesn't exhibit this behaviour, could be
```
Any reason not to do this now?

zerbina replied:
It felt like something unrelated to the rest of the PR, but you're right, there's not really a reason against it -- I'll change it here.

zerbina replied:
I've only changed the comment for now. The `__builtin_isnan` built-in is only available for GCC, Clang, and some other compilers aiming for GCC compatibility, but not for MSVC, and the only portable way of testing for NaNs that I've found would be inspecting the bit pattern (the C `isnan` function only became standard with C99).

Since `mChckNaN` is only intended for testing values resulting from arithmetic operations (which can never be signaling NaNs), I've changed the `mChckNaN` documentation to mention that signaling NaN behaviour is undefined.

saem reviewed (approved):
Quick question about not using the built-in now, but it's not really a blocker, feel free to merge.

zerbina committed:
ccgexprs: remove mention of `__builtin_isnan`

Apart from inspecting the bit patterns, there's no portable way of
testing for a NaN without triggering an FP exception in C89 (`isnan`
was added in C99) -- `__builtin_isnan` isn't available with MSVC.

zerbina committed:
ast_types: document signaling NaN behaviour

zerbina commented:
/merge
